File: samples/java/Bingo/Server/Play/src/main/java/systems/zlink/samples/bingo/server/play/infrastructure/zlink/spots/bingoroomspot/handlers/BingoRoomSettingsInitializer.java
package systems.zlink.samples.bingo.server.play.infrastructure.zlink.spots.bingoroomspot.handlers;

import systems.zlink.framework.messaging.ZLinkMessage;
import systems.zlink.framework.spots.ZLinkSpotCreateResponse;
import systems.zlink.samples.bingo.server.configuration.SampleTimings;
import systems.zlink.samples.bingo.server.play.domain.bingo.BingoRoomModels;
import systems.zlink.samples.bingo.server.play.infrastructure.zlink.spots.bingoroomspot.BingoRoomSpot;
import systems.zlink.samples.bingo.shared.contracts.Messages;

public final class BingoRoomSettingsInitializer {
    public ZLinkSpotCreateResponse handle(BingoRoomSpot spot, ZLinkMessage request) {
        spot.applySettings(decodeSettings(request));
        return ZLinkSpotCreateResponse.accept();
    }

    private BingoRoomModels.BingoRoomSettings decodeSettings(ZLinkMessage request) {
        if (request.isEmpty()) {
            return BingoRoomModels.BingoRoomSettings.create(
                    "two-player", 0, SampleTimings.DrawPeriod.toMillis());
        }
        Messages.BingoRoomCreateReq createRequest =
                request.decode(Messages.BingoRoomCreateReq.class);
        Messages.BingoRoomSettingsPayload settings = createRequest.getSettings();
        return new BingoRoomModels.BingoRoomSettings(
                settings.getRoomName(),
                settings.getMode(),
                settings.getRequiredPlayers(),
                settings.getMaxDrawNumber(),
                SampleTimings.DrawPeriod.toMillis(),
                settings.getPurpose(),
                settings.hasObservedRoomId() ? settings.getObservedRoomId() : null);
    }
}

File: samples/kotlin/Bingo/Server/Play/src/main/kotlin/systems/zlink/samples/kotlin/bingo/server/play/infrastructure/zlink/spots/bingoroomspot/handlers/BingoRoomSettingsInitializer.kt
package systems.zlink.samples.kotlin.bingo.server.play.infrastructure.zlink.spots.bingoroomspot.handlers

import systems.zlink.framework.messaging.ZLinkMessage
import systems.zlink.samples.kotlin.bingo.server.configuration.SampleTimings
import systems.zlink.samples.kotlin.bingo.server.play.domain.bingo.BingoRoomSettings
import systems.zlink.samples.kotlin.bingo.server.play.infrastructure.zlink.spots.bingoroomspot.BingoRoomSpot
import systems.zlink.samples.kotlin.bingo.shared.contracts.BingoRoomCreateReq

class BingoRoomSettingsInitializer {
    fun handle(spot: BingoRoomSpot, request: ZLinkMessage) {
        spot.applySettings(decodeSettings(request))
    }

    private fun decodeSettings(request: ZLinkMessage): BingoRoomSettings {
        if (request.isEmpty()) {
            return BingoRoomSettings.create("two-player", 0, SampleTimings.DrawPeriod.toMillis())
        }
        val settings = request.decode(BingoRoomCreateReq::class.java).settings
        return BingoRoomSettings(
            roomName = settings.roomName,
            mode = settings.mode,
            requiredPlayers = settings.requiredPlayers,
            maxDrawNumber = settings.maxDrawNumber,
            drawPeriodMillis = SampleTimings.DrawPeriod.toMillis(),
            purpose = settings.purpose,
            observedRoomId = if (settings.hasObservedRoomId()) settings.observedRoomId else null,
        )
    }
}
